satellite/accounting/tally/tally.go

```diff
@@ -38,204 +34,185 @@ type Config struct {
+	return &Observer{
+		Node:   make(map[storj.NodeID]float64),
+		Bucket: make(map[string]*accounting.BucketTally),
 	}
-	bucketReport(&totalTallies, "total")
+}
 
-	//store byte hours, not just bytes
-	numHours := time.Since(latestTally).Hours()
-	if latestTally.IsZero() {
-		numHours = 1.0 //todo: something more considered?
-	}
-	latestTally = time.Now().UTC()
+// ensureBucket returns bucket corresponding to the passed in path
+func (observer *Observer) ensureBucket(ctx context.Context, path metainfo.ScopedPath) *accounting.BucketTally {
+	bucketID := storj.JoinPaths(path.ProjectIDString, path.BucketName)
 
-	if len(nodeData) == 0 {
-		return latestTally, nodeData, bucketTallies, nil
-	}
-	for k := range nodeData {
-		nodeData[k] *= numHours //calculate byte hours
+	bucket, exists := observer.Bucket[bucketID]
+	if !exists {
+		bucket = &accounting.BucketTally{}
+		bucket.ProjectID = path.ProjectID
+		bucket.BucketName = []byte(path.BucketName)
+		observer.Bucket[bucketID] = bucket
 	}
-	return latestTally, nodeData, bucketTallies, err
+
+	return bucket
+}
+
+// Object is called for each object once.
+func (observer *Observer) Object(ctx context.Context, path metainfo.ScopedPath, pointer *pb.Pointer) (err error) {
+	bucket := observer.ensureBucket(ctx, path)
+	bucket.ObjectCount++
+	return nil
+}
+
+// InlineSegment is called for each inline segment.
+func (observer *Observer) InlineSegment(ctx context.Context, path metainfo.ScopedPath, pointer *pb.Pointer) (err error) {
+	bucket := observer.ensureBucket(ctx, path)
+	bucket.InlineSegments++
+	bucket.InlineBytes += int64(len(pointer.InlineSegment))
+	bucket.MetadataSize += int64(len(pointer.Metadata))
+
+	return nil
 }
 
-// bucketTallyAdd groups all the data based the passed pointer
-func bucketTallyAdd(s *accounting.BucketTally, pointer *pb.Pointer, last bool) {
-	switch pointer.GetType() {
-	case pb.Pointer_INLINE:
-		s.InlineSegments++
-		s.InlineBytes += int64(len(pointer.InlineSegment))
-		s.MetadataSize += int64(len(pointer.Metadata))
-
-	case pb.Pointer_REMOTE:
-		s.RemoteSegments++
-		s.RemoteBytes += pointer.GetSegmentSize()
-		s.MetadataSize += int64(len(pointer.Metadata))
+// RemoteSegment is called for each remote segment.
+func (observer *Observer) RemoteSegment(ctx context.Context, path metainfo.ScopedPath, pointer *pb.Pointer) (err error) {
+	bucket := observer.ensureBucket(ctx, path)
+	bucket.RemoteSegments++
+	bucket.RemoteBytes += pointer.GetSegmentSize()
+	bucket.MetadataSize += int64(len(pointer.Metadata))
+
+	// add node info
+	remote := pointer.GetRemote()
+	redundancy := remote.GetRedundancy()
+	segmentSize := pointer.GetSegmentSize()
+	minimumRequired := redundancy.GetMinReq()
+
+	if remote == nil || redundancy == nil || minimumRequired <= 0 {
+		// TODO: observer.log.Error("failed sanity check")
+		return nil
 	}
 
-	if last {
-		s.ObjectCount++
+	pieceSize := float64(segmentSize / int64(minimumRequired))
+	for _, piece := range pointer.GetRemote().GetRemotePieces() {
```

Comment:
you could say `range remote.GetRemotePieces()` instead

Reply:
We probably don't need to check, since there wouldn't be a panic when there aren't pices.